tests/modules/redshift_datasets/test_unit_redshift_connection_service.py

```diff
@@ -0,0 +1,249 @@
+from assertpy import assert_that
+from unittest.mock import MagicMock
+from .conftest import MockRedshiftClient, MockRedshiftDataClient, MockRedshiftServerlessClient
+from dataall.modules.redshift_datasets.services.redshift_connection_service import RedshiftConnectionService
+
+
+def test_create_redshift_serverless_connection(connection1_serverless):
+    # When connection is created
+    # Then
+    assert_that(connection1_serverless).is_not_none()
+    assert_that(connection1_serverless.connectionUri).is_not_none()
+    assert_that(connection1_serverless.redshiftType).is_equal_to('serverless')
+
+
+def test_create_redshift_cluster_connection(connection2_cluster):
+    # When connection is created
+    # Then
+    assert_that(connection2_cluster).is_not_none()
+    assert_that(connection2_cluster.connectionUri).is_not_none()
+    assert_that(connection2_cluster.redshiftType).is_equal_to('cluster')
+
+
+def test_create_redshift_connection_namespace_not_found(env_fixture, api_context_1, group, module_mocker):
+    mock_serverless = MagicMock()
+    module_mocker.patch(
+        'dataall.modules.redshift_datasets.services.redshift_connection_service.redshift_serverless_client',
+        return_value=mock_serverless,
+    )
+    mock_serverless.get_namespace_by_id.return_value = None
+
+    # Then
+    assert_that(RedshiftConnectionService.create_redshift_connection).raises(Exception).when_called_with(
+        uri=env_fixture.environmentUri,
+        admin_group=group.name,
+        data={
+            'connectionName': 'connection3',
+            'redshiftType': 'serverless',
+            'clusterId': None,
+            'nameSpaceId': 'not-existent-id',
+            'workgroup': 'workgroup-id',
+            'database': 'database_1',
+            'redshiftUser': None,
+            'secretArn': 'arn:aws:secretsmanager:*:111111111111:secret:secret-2',
+        },
+    ).contains('Redshift namespaceId not-existent-id does not exist')
+
+
+def test_create_redshift_connection_workgroup_not_in_namespace(env_fixture, api_context_1, group, module_mocker):
+    # Given a namespace but a workgroup that does not belong to it
+    mock_serverless = MagicMock()
+    module_mocker.patch(
```

Comment:
what's the lifetime of this patch? I suspect (i haven't used mocker fixture before) that because you use `module_mocker` instead of `mocker` this will not be undone after the test is done resulting in cross-condamination

Reply:
I have changed the mocks to be function-specific. And also for most fixtures for connections and datasets